File: tasks/2025/round2/homework.cpp
// @check-accepted: *
#include <algorithm>
#include <iostream>
using namespace std;

const int c = 550005;
long long n, m, t[c], h[c];  // feladatok szama, ido, feladat hossza, hatarido

long long order1[c], order2[c], inv2[c], spec[c];  // t illetve h szerint rendeze

bool cmp1(int a, int b) {
    return (t[a] < t[b]);
}
bool cmp2(int a, int b) {
    return (h[a] < h[b]);
}

long long fen[c], db[c];
long long lft[c], rght[c], mini[c], sum[c];

int lsb(int a) {
    return (a & -a);
}
void add(int a, int b, long long mul) {
    while (a <= n) {
        fen[a] += b * mul;
        db[a] += mul;
        a += lsb(a);
    }
}
long long ask(long long val) {
    int ans = 0, pos = 0;
    for (int i = 19; i >= 0; i--) {
        int kov = pos + (1 << i);
        if (kov <= n && fen[kov] <= val) {
            val -= fen[kov];
            pos = kov;
            ans += db[kov];
        }
    }
    return ans;
}
void range_add(int a, int l, int r, long long val) {
    if (lft[a] > r || rght[a] < l) {
        return;
    }
    if (l <= lft[a] && rght[a] <= r) {
        sum[a] += val;
        mini[a] += val;
        return;
    }
    int x = 2 * a, y = 2 * a + 1;
    range_add(x, l, r, val), range_add(y, l, r, val);
    mini[a] = sum[a] + min(mini[x], mini[y]);
}
int solve_nlogn() {
    for (int i = 1; i <= n; i++) {
        order1[i] = i, order2[i] = i;
    }
    sort(order1 + 1, order1 + n + 1, cmp1), sort(order2 + 1, order2 + n + 1, cmp2);
    for (int i = 1; i <= n; i++) {
        inv2[order2[i]] = i;
    }

    for (int i = 1; i <= n; i++) {
        long long val = t[order1[i]];
        add(i, val, 1);
    }

    int pos = 1;
    int po = 1;
    while (po < n) {
        po *= 2;
    }
    for (int i = po; i < 2 * po; i++) {
        lft[i] = i - po + 1, rght[i] = i - po + 1;
        if (lft[i] <= n) {
            sum[i] = h[order2[lft[i]]];
            mini[i] = sum[i];
        }
    }
    for (int i = po - 1; i >= 1; i--) {
        lft[i] = lft[2 * i], rght[i] = rght[2 * i + 1];
        mini[i] = min(mini[2 * i], mini[2 * i + 1]);
    }

    long long ans = ask(m);
    long long specsum = 0, specdb = 0;
    for (int i = 1; i <= n; i++) {
        int id = order1[i];
        add(i, t[id], -1);
        int hely = inv2[id];
        range_add(1, hely, n, -t[id]);
        specsum += t[id], specdb++;

        if (mini[1] >= 0) {
            ans = max(ans, 2 * specdb + ask(m - specsum));
        } else {
            range_add(1, hely, n, t[id]);
            specsum -= t[id], specdb--;
            add(i, t[id], 1);
        }
    }

    for (int i = 0; i < 2 * po; i++) {
        lft[i] = 0, rght[i] = 0, sum[i] = 0, mini[i] = 0;
    }
    for (int i = 0; i <= n; i++) {
        fen[i] = 0, db[i] = 0;
        order1[i] = 0, order2[i] = 0, inv2[i] = 0;
    }
    return ans;
}

signed main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    int tc;
    cin >> tc;
    for (int w = 1; w <= tc; w++) {
        cin >> n >> m;
        for (int i = 1; i <= n; i++) {
            cin >> t[i] >> h[i];
        }

        cout << solve_nlogn() << "\n";
    }
}

File: tasks/2025/round2/homework.py
#!/usr/bin/env python3
# TLE with pypy...
from sys import stdin
input = stdin.readline

def solve():
    def lsb(a):
        return a & -a

    def add(a, b, mul):
        while a <= n:
            fen[a] += b * mul
            db[a] += mul
            a += lsb(a)

    def add1(a, b):
        while a <= n:
            fen[a] += b 
            db[a] += 1
            a += lsb(a)
            
    def ask(val):
        ans = 0
        pos = 0
        for i in range(19, -1, -1):
            next_pos = pos + 2**i
            if next_pos <= n and fen[next_pos] <= val:
                val -= fen[next_pos]
                pos = next_pos
                ans += db[next_pos]
        return ans

    def range_add(a, l, r, val):
        if left[a] > r or right[a] < l:
            return
        if l <= left[a] <= right[a] <= r:
            summ[a] += val
            mini[a] += val
            return
        x = 2 * a
        y = x + 1
        range_add(x, l, r, val)
        range_add(y, l, r, val)
        mini[a] = summ[a] + min(mini[x], mini[y])

    input()
    n, m = map(int, input().split())
    t=[0] * (n+1)
    h=[0] * (n+1)
    for i in range(1, n + 1):
        t[i], h[i] = map(int, input().split())
        
    tmp=[i for i in range(n+1)]
    order1 = sorted(tmp, key=lambda x: t[x])
    order2 = sorted(tmp, key=lambda x: h[x])

    inv2=[0]*(n+1)
    for i in range(1, n + 1):
        inv2[order2[i]] = i

    fen = [0] * (n+1)
    db = [0] * (n+1)
    for i in range(1, n + 1):
        add1(i, t[order1[i]])

    po = 1
    while po < n:
        po *= 2
    left = [0] * (2*po)
    right = [0] * (2*po)
    mini = [0] * (2*po)
    summ = [0] * (2*po)
    for i in range(po, 2 * po):
        left[i] = right[i] = i - po + 1
        if left[i] <= n:
            summ[i] = h[order2[left[i]]]
            mini[i] = summ[i]

    for i in range(po - 1, 0, -1):
        left[i] = left[2 * i]
        right[i] = right[2 * i + 1]
        mini[i] = min(mini[2 * i], mini[2 * i + 1])

    ans = ask(m)
    specsum = specdb = 0

    for i in range(1, n + 1):
        id = order1[i]
        add(i, t[id], -1)
        pos = inv2[id]
        range_add(1, pos, n, -t[id])
        specsum += t[id]
        specdb += 1
        if mini[1] >= 0:
            ans = max(ans, 2 * specdb + ask(m - specsum))
        else:
            range_add(1, pos, n, t[id])
            specsum -= t[id]
            specdb -= 1
            add1(i, t[id])
    print(ans,flush=False)

tc = int(input())
for _ in range(tc):
    solve()
